lib/Infer/AbstractInterpreter.cpp

```diff
@@ -1,6 +1,268 @@
+      }
+
+      return Result;
+    }
+
+    llvm::KnownBits lshr(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      llvm::KnownBits Result(lhs.getBitWidth());
+      const auto width = Result.getBitWidth();
+
+      auto Op0KB = lhs;
+      if (rhs.isConstant()) {
+        auto Val = rhs.getConstant().getLimitedValue();
+        if (Val < 0 || Val >= width) {
+          return Result;
+        }
+        Op0KB.One.lshrInPlace(Val);
+        Op0KB.Zero.lshrInPlace(Val);
+        Op0KB.Zero.setHighBits(Val);
+        // setHighBits takes an unsigned int, so getLimitedValue is harmless
+        return Op0KB;
+      } else if (!rhs.isZero()) {
+	auto confirmedTrailingZeros = rhs.countMinTrailingZeros();
+	auto minNum = 1 << (confirmedTrailingZeros - 1);
+	// otherwise, it's poison
+	if (minNum < width)
+	  Result.Zero.setHighBits(minNum);
+      }
+
+      return Result;
+    }
+
+    llvm::KnownBits ashr(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      llvm::KnownBits Result(lhs.getBitWidth());
+      const auto width = Result.getBitWidth();
+
+      auto confirmedTrailingZeros = rhs.countMinTrailingZeros();
+      auto minNum = 1 << (confirmedTrailingZeros - 1);
+      if (lhs.One.isSignBitSet()) {
+	// confirmed: sign bit = 1
+	if (minNum < width)
+	  Result.One.setHighBits(minNum + 1);
+      } else if (lhs.Zero.isSignBitSet()) {
+	// confirmed: sign bit = 0
+	if (minNum < width)
+	  Result.Zero.setHighBits(minNum + 1);
+      }
+      return Result;
+    }
+
+    llvm::KnownBits eq(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      llvm::KnownBits Result(1);
+      if (lhs.isConstant() && rhs.isConstant() && (lhs.getConstant() == rhs.getConstant())) {
+	Result.One.setBit(0);
+	return Result;
+      }
+      if (((lhs.One & rhs.Zero) != 0) || ((lhs.Zero & rhs.One) != 0)) {
+	Result.Zero.setBit(0);
+	return Result;
+      }
+      return Result;
+    }
+
+    llvm::KnownBits ne(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      llvm::KnownBits Result(1);
+
+      if (lhs.isConstant() && rhs.isConstant() && (lhs.getConstant() == rhs.getConstant()))
+	Result.Zero.setBit(0);
+      if (((lhs.One & rhs.Zero) != 0) || ((lhs.Zero & rhs.One) != 0))
+	Result.One.setBit(0);
+      return Result;
+    }
+
+    llvm::KnownBits ult(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      llvm::KnownBits Result(1);
+      if (getUMax(lhs).ult(getUMin(rhs)))
+	Result.One.setBit(0);
+      if (getUMin(lhs).uge(getUMax(rhs)))
+	Result.Zero.setBit(0);
+      return Result;
+    }
+
+    llvm::KnownBits slt(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      llvm::KnownBits Result(1);
+      if (getSMax(lhs).slt(getSMin(rhs)))
+	Result.One.setBit(0);
+      if (getSMin(lhs).sge(getSMax(rhs)))
+	Result.Zero.setBit(0);
+      return Result;
+    }
+
+    llvm::KnownBits ule(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      llvm::KnownBits Result(1);
```

Comment:
instead of creating a tempoarary object and mutating it, you should just construct the value that you want in the first place, there are several ways to do it, I think any is fine

Reply:
return APInt(1, 0) or return APInt(1, 1) should work just fine

Reply:
We are not returning APInt here; we are returning KnownBits. KnownBits doesn't have any public constructor that can let you do that; hence, the reason for this thing. There's a private constructor with which we can accomplish something like this but I think that would be much more unreadable than the current form.

Reply:
ugh.